src/ore/src/channel.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License in the LICENSE file at the
+// root of this repository, or online at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Channel utilities and extensions.
+
+use std::collections::VecDeque;
+
+use async_trait::async_trait;
+
+/// Extensions for the receiving end of asynchronous channels.
+#[async_trait(?Send)]
+pub trait ReceiverExt<T> {
+    /// Receives all of the currently buffered elements on the channel, up to some max.
+    ///
+    /// This method returns `None` if the channel has been closed and there are no remaining
+    /// messages in the channel's buffer.
+    ///
+    /// If there are no messages in the channel's buffer, but the channel is not yet closed, this
+    /// method will sleep until a message is sent or the channel is closed. When woken it will
+    /// return up to max currently buffered elements.
+    async fn recv_many(&mut self, max: usize) -> Option<VecDeque<T>>;
```

Comment:
its unfortunate that the tokio receiver type doesnt expose its capacity :(

Can the docs mention that a good max is the channel's size? Additionally we should document that this is cancel-safe, and add a TODO to use `impl Iterator` when "impl trait in trait" is stabilized!

Reply:
> its unfortunate that the tokio receiver type doesnt expose its capacity :(

I know :(

I updated the docs to reflect all of this!

Also I changed the trait to be `Send`-able since when I tried using it inside of a `task::spawn(...)` I couldn't send the Future inside needed to be `Send`, since it might be moved to different threads between await points. This required `T` to be `Send`, which doesn't seem like too bad of a tradeoff. I _think_ we can solve this by using GATs, but that requires manually implementing `Future`, so for now I stuck with this